packages/signals/browser-signals/src/core/signal-generators/network.ts

```diff
@@ -0,0 +1,97 @@
+import { logger } from '../../lib/logger'
+import { SignalEmitter } from '../emitter'
+import { SignalGenerator } from './types'
+let origFetch: typeof window.fetch
+
+export function addFetchInterceptor(
+  onRequest: (rs: Parameters<typeof window.fetch>) => void,
+  onResponse: (rs: Response) => void
+) {
+  logger.debug('Adding fetch interceptor')
+  origFetch = window.fetch
+  window.fetch = async (...args) => {
+    try {
+      onRequest(args)
+    } catch (err) {
+      console.log('Error handling request: ', err)
+    }
+    const response = await origFetch(...args)
+    try {
+      onResponse(response.clone())
+    } catch (err) {
+      console.log('Error handling response: ', err)
+    }
+    return response
+  }
+}
+
+const normalizeHeaders = (headers: HeadersInit): Headers => {
+  return headers instanceof Headers ? headers : new Headers(headers)
+}
+
+const containsJSONContent = (headers: HeadersInit | undefined): boolean => {
+  if (!headers) {
+    return false
+  }
+  const normalizedHeaders = normalizeHeaders(headers)
+  return (
+    normalizedHeaders.get('content-type')?.includes('application/json') || false
+  )
+}
+
+export class NetworkGenerator implements SignalGenerator {
+  id = 'network'
+  register(emitter: SignalEmitter) {
+    const handleRequest = ([url, rq]: Parameters<typeof window.fetch>) => {
+      if (!rq || !rq.body) {
+        return
+      }
+      const rIsAbs = new RegExp('^(?:[a-z+]+:)?//', 'i')
+      if (
+        !url ||
+        (rIsAbs.test(url.toString()) &&
+          !url.toString().includes(window.location.hostname))
+      ) {
+        return
+      }
+
+      if (!containsJSONContent(rq.headers)) {
+        return
+      }
+
+      emitter.emit({
+        type: 'network',
+        data: {
+          action: 'Request',
+          url: url,
+          method: rq.method || '',
+          data: JSON.parse(rq.body.toString()),
+        },
+      })
+    }
+    const handleResponse = async (rs: Response) => {
+      if (!containsJSONContent(rs.headers)) {
+        return
+      }
+      const url = rs.url
+      if (!url || !new URL(url).hostname.includes(window.location.hostname)) {
```

Comment:
I'm not sure this works in the following scenario:

I'm on url: `app.segment.com` and I want to dispatch an event to `segment.com/api/foo=123`

- In addition, can we add (more) tests for this domain matching stuff to ensure that it only fires when we expect?

Reply:
Hmm, yeah, I think the right solution at that point would be a whitelist, which falls under the guiding principle of "Let the customer ask for it if they want it".  Separated out the logic 👍